diesel_cli/src/pretty_printing.rs

```diff
@@ -0,0 +1,194 @@
+
+
+/// Simple pretty printer hand tailored for the output generated
+/// by the `quote` - crate for schema inference
+///
+/// Rules:
+///   1. Seeing { increases indentation level
+///   2. Seeing } decreases indentation level
+///   3. Insert newline after {, }, ,, and ;
+///   4. Don't put spaces:
+///      between ident and !,
+///      between path segments and ::
+///      after ( and before )
+///      before ,
+pub fn format_schema<W>(schema: &str, mut output: W) where W: ::std::io::Write {
+    let mut out = String::new();
+    let mut indent = String::new();
+    let mut skip_space = false;
+    let mut last_char = ' ';
+    
+    for c in schema.chars() {
+        // quote inserts whitespaces at some strange location,
+        // remove them:        
+        match c {
+            '!' | '(' | ';' | ',' | '<' | ')' | '>' if last_char.is_whitespace()
+                => { out.pop(); }
+            ':' if last_char.is_whitespace() => {
+                out.pop();
+                match out.pop() {
+                    Some(c @ '>') => {
+                        out.push(c);
+                        out += " ";
+                    }
+                    Some(c) => {
+                        out.push(c);
+                    }
+                    _ => {}
+                }
+            }
+            _=> {}
+        }
+        if skip_space && c.is_whitespace() && last_char != '>' {
+            skip_space = false;
+            continue;
+        }
+        last_char = c;
+        skip_space = false;
+        
+        // there is already an empty line before {
+        // we need to remove the already inserted indent, because the
+        // new indent is smaller than the old one
+        if c == '}' {
+            while let Some(c) = out.pop() {
+                if c == '\n' {
+                    break;
+                }
+            }
+            indent.pop();
+            out += "\n";
+            out += &indent;
+        }
+        // push the current token to the output string
+        out = format!("{}{}", out, c);
+        
+        // we need to insert newlines in some places and adjust the indent
+        // also we need to remember if we could skip the next whitespace
+        match c {
+            ';' | ',' | '}'  => {
+                skip_space = true;
+                out += "\n";
+                out += &indent;
+            }
+            ':' | '(' => skip_space = true,
+            '{' => {
+                skip_space = true;
+                out += "\n";
+                indent += "\t";
+                out += &indent;
+            }
+            _=>{}
+        }        
+    }
+    write!(&mut output, "{}", out).unwrap();
+}
+
+
+#[cfg(test)]
+mod tests {
+    use super::format_schema;
+    use std::io::Cursor;
+    
```

Comment:
All these tests follow the same pattern. I think it'd be nice to reduce the boilerplate code a bit by adding a simple macro here:

```rust
macro_rules! test_format_output {
    ($test_name:ident, $input:expr => $output:expr) => {
        #[test]
        fn $test_name() {
            let out = Vec::<u8>::new();
            let mut c = Cursor::new(out);
            let input = $input;

            format_schema(input, &mut c);
            let format_output = String::from_utf8(c.into_inner()).unwrap();

            assert_eq!($output, format_output);
        }
    }
}
```

which you can then call like this:

```rust
test_format_output!{
    test_remove_whitespace_colon,
    ":: diesel :: types :: Text" => "::diesel::types::Text"
}
```